### DISCUSSION TITLE
github actions hides my jwt token retrieved from the backend

### DISCUSSION THREAD

**Discussion by haladamateusz** [community/community#37970] [Actions]
I have a CRON github action, which executes a python script.

First request goes well, the script gets jwt token from the backend. Then when I am trying to do the second request, guarded by JwtGuard, the token is not attached in Authentication header.

I am logging the headers on backend and 'Authentication' header seems to be 'undefined'.

When I am trying to display jwtToken obtained from backend with Auth request, in Github Actions I can only see '***'.

What is the proper way to manage the token in github actions and attach it to the next requests properly?

**Comment by airtower-luna** [comment]
> When I am trying to display jwtToken obtained from backend with Auth request, in Github Actions I can only see '***'.

This indicates your token was recognized as a secret (or maybe you added a mask for it?), so it's censored in the output. If you want to check it anyway, some options are:

* Output only the length of the token. No details, but a good plausibility check.
* If you really want to show the token (meaning anyone with log access can get it :warning:) you could base64 encode it or similar, so it's not the same thing to the masking system.

I'm afraid the rest of your question is impossible to answer without seeing some of your code, in particular how you get the token from where it's retrieved to where it should be used, and how the header is set there.

**Reply by haladamateusz** [reply]
After a short research, here is what I got:

The token is hidden in console output, but still present. Good catch from you.

Somehow, I still cannot attach the token to the request header. Here is my python code: 

```
            # returns token in json object
            self.TOKEN_REQUEST = requests.post(env.LOGIN_URL, env.CREDENTIALS)

            # outputs token length
            print(len(self.TOKEN_REQUEST.json().get('token')))

            # outputs token locally. Masks it in github actions
            print(self.TOKEN_REQUEST.json().get('token'))

            url = env.SAVE_INVOICE_LOG_URL + '/102022'
            token = self.TOKEN_REQUEST.json().get('token')
            headers = { 'Authorization': 'Bearer ' + token }

            self.INVOICE_EXISTS_REQUEST = requests.get(url, headers=headers)
```

Here is what I got in the backend logs. For the debugging purposes this route in not guarded by JwtGuard but the token is still attached (atleast it should be). Anyway Authorization header is missing:
```
Nov 1 01:33:57 PM   host:  {
Nov 1 01:33:57 PM    host: 'blabla',
Nov 1 01:33:57 PM    'user-agent': 'python-requests/2.28.1',
Nov 1 01:33:57 PM    accept: '*/*',
Nov 1 01:33:57 PM    'accept-encoding': 'blabla',
Nov 1 01:33:57 PM    'cdn-loop': 'cloudflare; subreqs=1',
Nov 1 01:33:57 PM    'cf-connecting-ip': 'blabla',
Nov 1 01:33:57 PM    'cf-ew-via': 'blabla',
Nov 1 01:33:57 PM    'cf-ipcountry': 'blabla',
Nov 1 01:33:57 PM    'cf-ray': 'blabla',
Nov 1 01:33:57 PM    'cf-visitor': '{"scheme":"https"}',
Nov 1 01:33:57 PM    'cf-worker': 'blabla',
Nov 1 01:33:57 PM    'true-client-ip': 'blabla',
Nov 1 01:33:57 PM    'x-forwarded-for': 'blabla, blabla',
Nov 1 01:33:57 PM    'x-forwarded-proto': 'https',
Nov 1 01:33:57 PM    'x-request-start': 'blabla'
```

When I try to run the same setup with postman/same script running from console and locally running backend, the Authorization header is properly set:
```
 host:  {
  authorization: 'Bearer eyJhbGci...',
  'user-agent': 'PostmanRuntime/7.29.2',
  accept: '*/*',
  'postman-token': 'blabla',
  host: 'localhost:3000',
  'accept-encoding': 'gzip, deflate, br',
  connection: 'keep-alive'
}
```
Looks like github actions don't allow me to add headers to the requests or I am adding them in a wrong way.

EDIT: I tried to deploy it to different CRON Jobs provider and the result is the same. Looks like there's a problem with my backend provider.

**Reply by airtower-luna** [reply]
Your code looks correct, as far as creating the `Authorization` header is concerned.

I notice that in your local sample the `host` header is `localhost:3000`. That seems relevant, the `requests` documentation [mentions that authorization headers will be removed under certain circumstances](https://requests.readthedocs.io/en/latest/user/quickstart/#custom-headers). One of them:

> Authorization headers will be removed if you get redirected off-host.

So if the `url` you use from Actions redirects to another host the authorization header won't make it there. Obviously I can't know if that's the case, but it seems plausible based on what you're seeing.

**Reply by haladamateusz** [reply]
Thank you for your help. I will try with another cloud provider. Everything worked on heroku but I have migrated to render. I will look for a different provider and see if it helps.

**Reply by Mayankm13** [reply]
@haladamateusz  can you please let me know if you find a soln for this issue, I am generating the  JWT token with python and try to use the token to clone the second repo

 But getting below error

`/usr/local/bin/docker exec  5987[18](https://github.oneeuronet.com/Euronet-REN/EuronetsoftwareRenaissanceOnline/runs/57255?check_suite_focus=true#step:6:19)1bf714065e04b80092f73ebe05708dc0b82f6f51de24ff78e72f90d4bc sh -c "cat /etc/*release | grep ^ID"
Error: Input required and not supplied: token`

`      - name: checkout renaissance-online-external-libs 
        uses: actions/checkout@v3
        with:
          repository: Mayank/renaissance-online-external-libs
          ref: master
          path: renaissance-online-external-libs
          token: ${{ env.TOKEN }}  `

**Reply by haladamateusz** [reply]
@Mayankm13 
> @haladamateusz can you please let me know if you find a soln for this issue, I am generating the JWT token with python and try to use the token to clone the second repo
> 
> But getting below error
> 
> `/usr/local/bin/docker exec 5987[18](https://github.oneeuronet.com/Euronet-REN/EuronetsoftwareRenaissanceOnline/runs/57255?check_suite_focus=true#step:6:19)1bf714065e04b80092f73ebe05708dc0b82f6f51de24ff78e72f90d4bc sh -c "cat /etc/*release | grep ^ID" Error: Input required and not supplied: token`
> 
> ` - name: checkout renaissance-online-external-libs uses: actions/checkout@v3 with: repository: Mayank/renaissance-online-external-libs ref: master path: renaissance-online-external-libs token: ${{ env.TOKEN }} `

The problem was not related to cloud provider but to domain settings. I changed https://my-website.com to https://www.my-website.com and everything works like a charm.

In your case, it looks like you are not attaching the token properly to the request.
